javalib/src/main/scala/java/net/URLDecoder.scala

```diff
@@ -0,0 +1,84 @@
+package java.net
+
+// Credits to Scala.js
+
+import java.io.UnsupportedEncodingException
+import java.nio.{CharBuffer, ByteBuffer}
+import java.nio.charset.{Charset, MalformedInputException}
+
+object URLDecoder {
+
+  @Deprecated
+  def decode(s: String): String = decodeImpl(s, Charset.defaultCharset())
+
+  def decode(s: String, enc: String): String = {
+    /* An exception is thrown only if the
+     * character encoding needs to be consulted
+     */
+    lazy val charset = {
+      if (!Charset.isSupported(enc))
+        throw new UnsupportedEncodingException(enc)
+      else
+        Charset.forName(enc)
+    }
+
+    decodeImpl(s, charset)
+  }
+
+  private def decodeImpl(s: String, charset: => Charset): String = {
```

Comment:
This decoder is too slow. I've already provided a fast path patch for similar http4s decoder that saves more than 5% CPU cycles on our services: https://github.com/http4s/http4s/pull/5556

Reply:
Thanks @plokhotnyuk, I'll swap implementations